src/cascadia/TerminalControl/TermControl.cpp

```diff
@@ -227,34 +253,180 @@ namespace winrt::Microsoft::Terminal::TerminalControl::implementation
+    // Return Value:
+    // - <none>
+    void TermControl::_SelectSearchResult(std::optional<bool> goForward)
+    {
+        if (_searchState.has_value() && _searchState.value().Matches.has_value())
+        {
+            auto& state = _searchState.value();
+            auto& matches = state.Matches.value();
+
+            if (goForward.has_value())
+            {
+                state.UpdateIndex(goForward.value());
+
+                const auto currentMatch = state.GetCurrentMatch();
+                if (currentMatch.has_value())
+                {
+                    auto lock = _terminal->LockForWriting();
+                    _terminal->SetBlockSelection(false);
+                    _terminal->SelectNewRegion(currentMatch.value().first, currentMatch.value().second);
+                    _renderer->TriggerSelection();
+                }
+            }
+
+            if (_searchBox)
+            {
+                _searchBox->SetStatus(gsl::narrow<int32_t>(matches.size()), state.CurrentMatchIndex);
+                _searchBox->SetNavigationEnabled(!_searchState.value().Matches.value().empty());
+            }
+        }
+    }
+
+    // Method Description:
+    // - Search for a single value in a background
+    // Arguments:
+    // - search: search object to use to find the next match
+    // Return Value:
+    // - <none>
+    winrt::Windows::Foundation::IAsyncOperation<bool> TermControl::_SearchOne(Search& search)
+    {
+        bool found{ false };
+        auto weakThis{ this->get_weak() };
+
+        co_await winrt::resume_background();
+        if (auto control{ weakThis.get() })
+        {
+            // We don't lock the terminal for the duration of the entire search,
+            // since if the terminal was modified the search ID will be updated.
+            auto lock = _terminal->LockForWriting();
+            found = search.FindNext();
+        }
+
+        co_await winrt::resume_foreground(Dispatcher());
+        co_return found;
+    }
+
+    // Method Description:
+    // - The handler for the "search criteria changed" event. Clears selection and initiates a new search.
+    // Arguments:
+    // - text: the text to search
+    // - goForward: indicates whether the search should be performed forward (if set to true) or backward
+    // - caseSensitive: boolean that represents if the current search is case sensitive
+    // Return Value:
+    // - <none>
+    void TermControl::_SearchChanged(const winrt::hstring& text, const bool goForward, const bool caseSensitive)
+    {
+        if (_searchBox && _searchBox->Visibility() == Visibility::Visible)
+        {
+            // Clear the selection reset the anchor
+            _terminal->ClearSelection();
             _renderer->TriggerSelection();
+
+            const SearchState searchState{ text, caseSensitive };
+            _searchState.emplace(searchState);
+            _SearchAsync(goForward, SearchAfterChangeDelay);

```

Comment:
Careful- this is directly hooked up to the search textbox change handler, right? It will enqueue one search for every character typed, and then all of them will wait `delay` and then abort due to the IDs not matching. Ideally, we'd put _this_ in a throttled function as well, so that it only gets _invoked_ once per `delay` rather than invoking N times and noping out after `delay`.